File: selenium/src/test/java/com/qacommercelab/pages/ProductsPage.java
package com.qacommercelab.pages;

import com.qacommercelab.utils.ConfigReader;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;

public class ProductsPage {

    private final WebDriver driver;
    private final WebDriverWait wait;

    private final By searchInput = By.cssSelector("[data-testid='product-search']");

    private final By categorySelect = By.id("c");

    private final By sortSelect = By.id("o");

    private final By resultCount = By.id("cnt");

    private final By emptyProducts = By.cssSelector("[data-testid='empty-products']");

    private final By cartCount = By.cssSelector("[data-testid='cart-count']");

    public ProductsPage(WebDriver driver) {
        this.driver = driver;

        this.wait = new WebDriverWait(
                driver,
                Duration.ofSeconds(
                        ConfigReader.getInt("explicitWait")));
    }

    public ProductsPage open() {
        driver.get(ConfigReader.get("baseUrl") + "/products");

        wait.until(
                ExpectedConditions.visibilityOfElementLocated(searchInput));

        return this;
    }

    public void search(String text) {
        WebElement input = wait.until(
                ExpectedConditions.visibilityOfElementLocated(searchInput));

        input.clear();
        input.sendKeys(text);
    }

    public void selectCategory(String category) {
        Select select = new Select(
                wait.until(
                        ExpectedConditions.visibilityOfElementLocated(
                                categorySelect)));

        select.selectByVisibleText(category);
    }

    public void selectSort(String sortOption) {
        Select select = new Select(
                wait.until(
                        ExpectedConditions.visibilityOfElementLocated(
                                sortSelect)));

        select.selectByVisibleText(sortOption);
    }

    public String getResultCountText() {
        return wait.until(
                ExpectedConditions.visibilityOfElementLocated(resultCount)).getText();
    }

    public int getVisibleProductCount() {
        return driver.findElements(
                By.cssSelector("[data-testid^='product-card-']")).size();
    }

    public boolean isEmptyStateDisplayed() {
        return wait.until(
                ExpectedConditions.visibilityOfElementLocated(emptyProducts)).isDisplayed();
    }

    public boolean isProductDisplayed(int productId) {
        return wait.until(
                ExpectedConditions.visibilityOfElementLocated(
                        By.cssSelector(
                                "[data-testid='product-card-" +
                                        productId +
                                        "']")))
                .isDisplayed();
    }

    public String getProductCardText(int productId) {
        return wait.until(
                ExpectedConditions.visibilityOfElementLocated(
                        By.cssSelector(
                                "[data-testid='product-card-" +
                                        productId +
                                        "']")))
                .getText();
    }

    public boolean isAddToCartEnabled(int productId) {
        return wait.until(
                ExpectedConditions.presenceOfElementLocated(
                        By.cssSelector(
                                "[data-testid='add-cart-" +
                                        productId +
                                        "']")))
                .isEnabled();
    }

    public void addProductToCart(int productId) {
        wait.until(
                ExpectedConditions.elementToBeClickable(
                        By.cssSelector(
                                "[data-testid='add-cart-" +
                                        productId +
                                        "']")))
                .click();
    }

    public int getCartCount() {
        String text = wait.until(
                ExpectedConditions.visibilityOfElementLocated(cartCount)).getText();

        return Integer.parseInt(text);
    }

    public List<Double> getVisiblePrices() {
        return driver.findElements(
                By.cssSelector(
                        "[data-testid^='product-card-'] .price"))
                .stream()
                .map(WebElement::getText)
                .map(this::parsePrice)
                .toList();
    }

    private double parsePrice(String text) {
        return Double.parseDouble(
                text
                        .replace("$", "")
                        .replace(",", "")
                        .trim());
    }
}

File: playwright/pages/ProductsPage.ts
import { expect, Locator, Page } from "@playwright/test";

export class ProductsPage {
  readonly page: Page;

  readonly searchInput: Locator;
  readonly categorySelect: Locator;
  readonly sortSelect: Locator;
  readonly resultCount: Locator;
  readonly emptyProducts: Locator;
  readonly cartCount: Locator;
  readonly productCards: Locator;

  constructor(page: Page) {
    this.page = page;

    this.searchInput = page.getByTestId("product-search");
    this.categorySelect = page.locator("#c");
    this.sortSelect = page.locator("#o");
    this.resultCount = page.locator("#cnt");

    this.emptyProducts = page.getByTestId("empty-products");
    this.cartCount = page.getByTestId("cart-count");

    this.productCards = page.locator("[data-testid^='product-card-']");
  }

  async open(): Promise<void> {
    await this.page.goto("/products");

    await expect(this.searchInput).toBeVisible();
  }

  productCard(productId: number): Locator {
    return this.page.getByTestId(`product-card-${productId}`);
  }

  addButton(productId: number): Locator {
    return this.page.getByTestId(`add-cart-${productId}`);
  }

  async search(value: string): Promise<void> {
    await this.searchInput.fill(value);
  }

  async selectCategory(category: string): Promise<void> {
    await this.categorySelect.selectOption({
      label: category,
    });
  }

  async selectSort(option: string): Promise<void> {
    await this.sortSelect.selectOption({
      label: option,
    });
  }

  async getVisibleProductCount(): Promise<number> {
    return this.productCards.count();
  }

  async expectProductVisible(productId: number): Promise<void> {
    await expect(this.productCard(productId)).toBeVisible();
  }

  async expectProductHidden(productId: number): Promise<void> {
    await expect(this.productCard(productId)).toBeHidden();
  }

  async getProductText(productId: number): Promise<string> {
    return (await this.productCard(productId).textContent())?.trim() ?? "";
  }

  async addProduct(productId: number): Promise<void> {
    await this.addButton(productId).click();
  }

  async getCartCount(): Promise<number> {
    const text = (await this.cartCount.textContent())?.trim() ?? "0";

    return Number(text);
  }

  async getVisiblePrices(): Promise<number[]> {
    const cards = this.productCards;

    const count = await cards.count();

    const prices: number[] = [];

    for (let index = 0; index < count; index++) {
      const text = (await cards.nth(index).textContent()) ?? "";

      const match = text.match(/\$([\d,]+(?:\.\d{1,2})?)/);

      if (!match) {
        throw new Error(`Price not found in product card: ${text}`);
      }

      prices.push(Number(match[1].replace(/,/g, "")));
    }

    return prices;
  }
}
